### PR TITLE
Checkpoint conversion tools

### PR DESCRIPTION
**Pull request by tjruwase**
Tools for converting checkpoints.

### CODE REVIEW

**Comment by tjruwase** [comment]
@stas00 FYI

**Comment by stas00** [comment]
OK, to complete the conversion the missing bits are:

```
sd["args"].tensor_model_parallel_size = 1
sd["args"].pipeline_model_parallel_size = 1
sd["args"].consumed_train_samples = 0
sd["args"].consumed_valid_samples = 0
```

the first 2 of course need to be adjusted to the target tp/pp sizes.

the last 2 need to be reset otherwise meg tries to resume training from some really high number of a sample which should be 0 instead.

---

I haven't quite figured out how to solve the `padded_vocab_size` being larger than the vocab. Probably needs to be truncated to the vocab size before saving the embeddings. 

the workaround is to use the actual padded vocab size when finetuning, i.e.:

```
--make-vocab-size-divisible-by 50688 
```
for when the padded vocab ends up being 50688.

**Comment by stas00** [comment]
Then the files layout, clearly Meg-LM expects this layout:
```
iter_0000001/mp_rank_00_000/model_optim_rng.pt
latest_checkpointed_iteration.txt
```
whereas in the Meg-DS tree it wants:
```
iter_0000001/mp_rank_00/model_optim_rng.pt
latest_checkpointed_iteration.txt
```
no `_000` at the end.

we could probably convert `global_step37876` to `iter_0037876` to help the user know which iteration the training is coming from.

**Comment by tjruwase** [comment]
> the first 2 of course need to be adjusted to the target tp/pp sizes.
> 
> the last 2 need to be reset otherwise meg tries to resume training from some really high number of a sample which should be 0 instead.

Thanks for the feedback. I can update deepspeed_to_megatron.py for the first 2 to fix an inconsistency in the checkpoint state. However, I am unsure where to handle the last 2 since it would prevent the converted checkpoint from being used for continued training. So perhaps the finetuning script should handle the last 2. What do you think?

**Comment by stas00** [comment]
but this checkpoint can't be loaded for continued training at the moment. e.g. it lacks the `iteration` entry, and Meg crashes w/o `--finetune` because of that.

I'm not sure how you'd change the `finetune` script to ignore `consumed_train_samples` because it also checkpoints and should be able to resume from its `consumed_train_samples` ongoing record.

Perhaps we have 2 different modes here:

1. reshape the checkpoint, but presume continued training
2. convert for a release purpose, resetting some counters.

**Comment by stas00** [comment]
Found one more culprit - remember how Jared's script was asking for a megatron clone path when doing the conversion? It proved to be essential, since if we use the default Meg-DS when converting, it then fails to `torch.load` when attempting to use Meg-LM:

```
Traceback (most recent call last):
  File "pretrain_gpt.py", line 124, in <module>
    pretrain(train_valid_test_datasets_provider, model_provider, forward_step,
  File "/mnt/nvme1/code/huggingface/Megatron-LM/megatron/training.py", line 112, in pretrain
    model, optimizer, lr_scheduler = setup_model_and_optimizer(model_provider)
  File "/mnt/nvme1/code/huggingface/Megatron-LM/megatron/training.py", line 325, in setup_model_and_optimizer
    args.iteration = load_checkpoint(model, optimizer, lr_scheduler)
  File "/mnt/nvme1/code/huggingface/Megatron-LM/megatron/checkpointing.py", line 314, in load_checkpoint
    state_dict = torch.load(checkpoint_name, map_location='cpu')
  File "/home/stas/anaconda3/envs/py38-pt19/lib/python3.8/site-packages/torch/serialization.py", line 607, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
  File "/home/stas/anaconda3/envs/py38-pt19/lib/python3.8/site-packages/torch/serialization.py", line 882, in _load
    result = unpickler.load()
  File "/home/stas/anaconda3/envs/py38-pt19/lib/python3.8/site-packages/torch/serialization.py", line 875, in find_class
    return super().find_class(mod_name, name)
ModuleNotFoundError: No module named 'megatron.enums'
``` 

Things have changed in Meg-DS and it now it can't find `megatron.enums`. In Meg-LM this is `megatron.model.enums` instead.

**Comment by tjruwase** [comment]
> Additionally, we could probably convert `global_step37876` to `iter_0037876` to help the user know which iteration the training is coming from. Rather than iter=1, incidentally you can then save it in the dict as `iteration=37876` which it wants anyway for non-finetune.
Are you seeing iter=1 in the converted checkpoint?

**Comment by stas00** [comment]
I think it perhaps expects the top-level `iteration` key and not `args.iteration`? remember how I mentioned there were 3 keys it expects? So I think it wants:

```
ITERATION_KEY = 'iteration'
...
    checkpoint_sd[ITERATION_KEY] = iteration
```    

in your code

File: tools/convert_checkpoint/deepspeed_to_megatron.py [hunk 1]
```diff
@@ -21,6 +21,7 @@ def parse_arguments():
     parser.add_argument('--output_folder', default=None, type=str, help='Output Megatron checkpoint folder')
     parser.add_argument('--target_tp', default=None, type=int, help='Target TP degree')
     parser.add_argument('--target_pp', default=None, type=int, help='Target PP degree')
+    parser.add_argument('--for_release', action='store_true', help='Convert for release purpose, reset some (progress) counters.')
```
Comment (stas00) [inline tools/convert_checkpoint/deepspeed_to_megatron.py:24]: add let's add this to README as well

**Comment by tjruwase** [comment]
> think it perhaps expects the top-level `iteration` key and not `args.iteration`? remember how I mentioned there were 3 keys it expects? So I think it wants:

It would be great to get some clarity on which iteration to use here. 

Also should it be 

1. ```python checkpoint_sd['args'][ITERATION_KEY]``` or 
2. ```python checkpoint_sd[ITERATION_KEY]```?

**Comment by stas00** [comment]
> > think it perhaps expects the top-level `iteration` key and not `args.iteration`? remember how I mentioned there were 3 keys it expects? So I think it wants:
> 
> It would be great to get some clarity on which iteration to use here.

the problem is that it seems that Meg-DS doesn't update Meg's native `iteration` variable soon enough and so the saved checkpoint reports the outdated lower iteration. So most likely the fix is needed inside Meg-DS to at the very least sync Meg's native `iteration` variable with Meg-DS value of the same.

> Also should it be
> 
>   1. `python checkpoint_sd['args'][ITERATION_KEY]` or
>   2. `python checkpoint_sd[ITERATION_KEY]`?

The latter. 

But it seems to be pointless, because once I add the missing key, it then fails with:

```
 loading checkpoint from /hf/Megatron-DeepSpeed-master/data/1B3-PP4-TP4-Meg at iteration 1
 checkpoint version 3.0
Unable to load optimizer from checkpoint /hf/Megatron-DeepSpeed-master/data/1B3-PP4-TP4-Meg/iter_0000001/mp_rank_00/model_optim_rng.pt. Specify --no-load-optim or --finetune to prevent attempting to load the optimizer state, exiting ...
```

we can't manifest optimizer states for Meg-LM out of nowhere, so it appears that after the conversion only inference or finetuning is possible. In which case it's probably pointless to try to keep `consumed_train_samples` and its friend.

Perhaps let's for now handle just the clear case of inference/finetuning with an assumption that finetuning will require a different dataset?

**Comment by stas00** [comment]
So I think the only remaining thing to address (other than embeddings) is:
https://github.com/microsoft/Megatron-DeepSpeed/pull/14#issuecomment-917211393

**Comment by stas00** [comment]
In another checkpoint I was given the discrepancy is 2x,

Meg-DS file is `global_step37876` but `args.iteration=18931` which is ~0.5 of the former. Weird. Looks like the 2 counters aren't in sync at all.

**Comment by stas00** [comment]
OK, I have figured this one out. You were getting the wrong `iteration`, you need this one:
```
f = "global_step37876/mp_rank_00_model_states.pt"
sd = torch.load(f)
sd["iteration"]
```

this is the real iteration, but the original code args.iteration is whatever was the iteration at the start of training. Does it make sense?

**Comment by stas00** [comment]
@tjruwase, 2 more things that I discovered are different from the checkpoint generated by Meg-LM natively.

```
- embeddings["word_embeddings.weight"]
+ embeddings["word_embeddings"]["weight"]
```
and:
```
+ embeddings["position_embeddings.weight"]
-  embeddings["position_embeddings"]["weight"]
```

So I think after this fix, the resulting checkpoint will be matching the native one.

Plus the resulting file structure: https://github.com/microsoft/Megatron-DeepSpeed/pull/14#issuecomment-917289494

and then it's good to be merged.

I did multiple tests on the final stage meg2hf and the conversion appears to be correct.

**Comment by stas00** [comment]
This is good to merge, @tjruwase! Thank you!

**Comment by stas00** [comment]
@tjruwase, I made a small change to your work to separate the creation of the checkpoint and saving it, so that I could re-use it create the HF transformers checkpoint on the fly.

Also made the PP/TP size 1 by default, since in the HF case it's always that for now.

If it looks acceptable to you perhaps let's merge this back into your master tree?
https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/110/commits/b778da32fec0748c03cc4787aee3389eaee1f7be

**Comment by stas00** [comment]
here you go - I added 2 more where I made the scripts executable ;)

```
git checkout -b convert-meg-ds-to-hf
git remote add other https://github.com/bigscience-workshop/Megatron-DeepSpeed
git fetch other
git cherry-pick bea5ded
git cherry-pick d6c2a80
git cherry-pick 26f18b5
git cherry-pick 2f662e8
git push --set-upstream origin convert-meg-ds-to-hf
```

**Pull request merged**